GTScript-Parallel-model.md

```diff
@@ -134,17 +180,233 @@ The computation domain of every statement is extended to ensure that any require
+- Inside the if and else blocks the same field cannot be written to **and** read with an offset in the parallel axises (order does not matter).
+
+**Example for conditionals on field expressions**
+
+```python
+with computation():
+    with interval(...):
+        if field:
+            a = 1
+            b = 2
+        else:
+            a = 2
+            b = 1
+```
+
+translates to:
+
 ```python
 for k in range(start, end):
-    parfor [i_start-2:i_end+1, j_start-1:j_end+2]:
-        u[i, j, k] = 1
-    parfor [i_start:i_end, j_start:j_end]:
-        b[i, j, k] = u[i-2,j,k] + u[i+1,j,k] + u[i,j-1,k] + u[i,j-2,k]
+    parfor ij:
+        mask[i,j] = field[i,j,k]
+    parfor ij:
+        if mask[i,j]:
+            a[i, j, k] = 1
+    parfor ij:
+        if mask[i,j]:
+            b[i, j, k] = 2
+    parfor ij:
+        if not mask[i,j]:
+            a[i, j, k] = 2
+    parfor ij:
+        if not mask[i,j]:
+            b[i, j, k] = 1
+```
+
+The following cases are illegal:
+
+```python
+with computation():
+    with interval(...):
+        if field:
+            b = a[1, 0, 0] # read with offset in 'I' from updated field 'a'
+            a = 1
+            c = a[0, 1, 0] # read with offset in 'J' from updated field 'a'
+
+with computation():
+    with interval(...):
+        if field:
+            a = 1
+        else:
+            b = a[1, 0, 0] # read with offset in 'I' from updated field 'a'
+
+with computation(...):
+    with interval(...):
+        if field:
+            a = a[0, 1, 0] # self assignment with offset (i.e. a read with offset and write)
+```
+
+## Off-center writes in `K`
+
+The vertical direction, `K`, is special in weather and clinate applicaitons. It is used nut just for stencil computations but also for implicit skemes and other algorthms in physical parametrization.
+This is the reason why the vertical look is kept separate from tne horizontal IJ look in the parallel model.
+
+The vertical loop can have _directions_, indicated by the iteraiton policies FORWARD, BACKWARD, or can left unspecified, intending that it could be possibly parallelized. To implement certain skemes it is natural to allow for off-center writes in the vertical direction, since the programmer can know in which order the values are accessed.
+
+There are different uses in which writing off-center can work with clear semantics: write at a location before the iteration loop reaches it (e.g., offset greater than zero in a FORWARD loop), and write at a location after the iteration loop reaches it (e.g., offset less than zero in a FORWARD loop). Both cases are valid. In the first we want to make the values visible in the same computation, in the second case we want to make the values available for subsequent computations. There are cases with conditionals in which the two cases above can be mixed up, but they are deterministic and creates no problems, even though understanding the code by the programmers can be difficult.
```

Comment:
```suggestion
There are different use cases in which writing off-center can work with clear semantics: write at a K-index before the iteration loop reaches it (e.g., offset greater than zero in a FORWARD loop), and write at a location after the iteration loop reaches it (e.g., offset less than zero in a FORWARD loop). Both cases are valid. In the first we want to make the values visible in the same computation, in the second case we want to make the values available for subsequent computations. There are cases with conditionals in which the two cases above can be mixed, but they must be deterministic (no read-before-write) and the resulting code can be hard to understand.
```

Reply:
About the read-before write and the "must be". The computation, according to the parallel model is always deterministic. The result could be visible in the next k-iteration or not, but it is always visible when the computation completes, which means you can use the off-center write to prepare the data for a subsequent computation. So I think the "must be" is not correct and the concern with read-before-write comment unnecessary.